lib/fetch-server-test.js

```diff
@@ -0,0 +1,155 @@
+import isFunction from 'lodash/isFunction';
+import intersection from 'lodash/intersection';
+import difference from 'lodash/difference';
+
+import * as sinon from 'sinon';
+
+const VALID_METHODS = [`GET`, `POST`, `PATCH`, `PUT`, `DELETE`];
+
+/**
+ * Utility class to stub fetch requests with sinon. Instantiate with a set of
+ * route handlers mapping URL (sub)strings to canned response values, or functions
+ * which return canned responses/response values. Values will be serialized as JSON
+ * and returned in a Fetch-like Promise object.
+ *
+ * @example
+ * const fetchServer = new FetchServer({
+ *   '/foo.json': 'bar',
+ *   '/baz/quux': (url, params) => !!params.foo ? 'a' : 'z',
+ * }, {debug: true});
+ * fetchServer.start();
+ * const res = await fetch('https://myserver/foo.json');
+ * const val = await res.json(); // 'bar'
+ * 
+ * @example 
+ * You can also have the following handler types where key is any of the HTTP methods
+ * const handler = {
+ *   '/foo/bar`: {
+ *      GET: {
+ *        'content-type': 'application/json',
+ *        'status': 200,
+ *        'result': {}
+ *      },
+ *      // Takes a non-async function, and creates Promise
+ *      POST: (url, params) => ({
+ *        'content-type': 'text/html',
+ *        'status': 200,
+ *        'result': {}
+ *      })
+ *   }
+ * }
+ */
+export class FetchServer {
+  constructor(handlers, { debug = false } = {}) {
+    this.handlers = handlers;
+    this.debug = debug;
+    this.handle = this.handle.bind(this);
+  }
+
+  handle(url, params) {
+    // default method GET
+    const HTTPmethod = params.method || `GET`;
+    let response;
+    for (const [route, handler] of Object.entries(this.handlers)) {
+
+      // Block 1 : parses body
+      if (url.includes(route)) {
+        if (params && params.body) {
+          try {
+            params = JSON.parse(params.body);
+          } catch (e) {
+            if (e instanceof SyntaxError) {
+              // not JSON parse-able, must be a string
+              params = params.body;
+            } else {
+              throw new Error(`Invalid body: ${params.body}`);
+            }
+          }
+        }
+
+        // handler is either normal or async function. 
+        if (isFunction(handler)) response = handler(url, params);
```

Comment:
So if `handler` is already async, `response` will be a promise? If not it will be wrapped in promise below? This makes sense, just want to make sure I understand - a comment describing this feature here would be very helpful to readers.

Reply:
Would be great to restructure this somehow to get this line closer to the promise-wrapping logic, maybe even just reversing this if-condition.

One more nit - mandatory brackets with if statements is pretty strict style at MP, guarantee you'll run into pushback from others if we don't fix it here :)

Reply:
right yes will make sure to update all the styles.